python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+    def partition_base_dir(self):
+        return frombytes(self.options.partition_base_dir)
+
+    @partition_base_dir.setter
+    def partition_base_dir(self, value):
+        self.options.partition_base_dir = tobytes(value)
+
+    @property
+    def exclude_invalid_files(self):
+        return self.options.exclude_invalid_files
+
+    @exclude_invalid_files.setter
+    def exclude_invalid_files(self, bint value):
+        self.options.exclude_invalid_files = value
+
+    @property
+    def ignore_prefixes(self):
+        return [frombytes(p) for p in self.options.ignore_prefixes]
+
+    @ignore_prefixes.setter
+    def ignore_prefixes(self, values):
+        self.options.ignore_prefixes = [tobytes(v) for v in values]
+
+
+cdef class DataSourceDiscovery:
+
+    cdef:
+        shared_ptr[CDataSourceDiscovery] wrapped
+        CDataSourceDiscovery* discovery
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef init(self, shared_ptr[CDataSourceDiscovery]& sp):
+        self.wrapped = sp
+        self.discovery = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CDataSourceDiscovery]& sp):
+        cdef DataSourceDiscovery self = \
+            DataSourceDiscovery.__new__(DataSourceDiscovery)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CDataSourceDiscovery] unwrap(self):
+        return self.wrapped
+
+    @property
+    def schema(self):
+        cdef shared_ptr[CSchema] schema = self.discovery.schema()
+        if schema.get() == nullptr:
+            return None
+        else:
+            return pyarrow_wrap_schema(schema)
+
+    @property
+    def partition_scheme(self):
+        cdef shared_ptr[CPartitionScheme] scheme
+        scheme = self.discovery.partition_scheme()
+        if scheme.get() == nullptr:
+            return None
+        else:
+            return PartitionScheme.wrap(scheme)
+
+    @partition_scheme.setter
+    def partition_scheme(self, PartitionScheme scheme not None):
+        check_status(self.discovery.SetPartitionScheme(scheme.unwrap()))
+
+    @property
+    def root_partition(self):
+        cdef shared_ptr[CExpression] expr = self.discovery.root_partition()
+        if expr.get() == nullptr:
+            return None
+        else:
+            return Expression.wrap(expr)
+
+    def inspect(self):
+        cdef CResult[shared_ptr[CSchema]] result
+        with nogil:
+            result = self.discovery.Inspect()
+        return pyarrow_wrap_schema(GetResultValue(result))
+
+    def finish(self):
+        cdef CResult[shared_ptr[CDataSource]] result
+        with nogil:
+            result = self.discovery.Finish()
+        return DataSource.wrap(GetResultValue(result))
+
+
+cdef class FileSystemDataSourceDiscovery(DataSourceDiscovery):
+
+    cdef:
+        CFileSystemDataSourceDiscovery* filesystem_discovery
+
+    def __init__(self, FileSystem filesystem not None,
+                 Selector selector not None, FileFormat format not None,
```

Comment:
Can you allow a list of FileStats here as well? 

(there is an alternative FileSystemDataSourceDiscovery constructor that accepts a vector of FileStats)

Reply:
I concur, this is an important feature because selector can't express all filters (and likely never will). This constructor exists as a convenience to crawl a directory.

Reply:
Added support for passing either file *paths* or a file selector. Passing FileStats objects seems a pretty weird, what if the passed FileStats objects are originating from a different filesystem?

Reply:
I've removed the tasks so we directly yield batches now.